File: src/backend/app/models/subtitle.py
from typing import Optional

from pydantic import BaseModel, model_validator


class SubtitleCue(BaseModel):
    index: int
    start: float
    end: float
    source_text: str
    target_text: str
    display_text: Optional[str] = None

    @model_validator(mode="after")
    def build_display_text(self) -> "SubtitleCue":
        if self.display_text is None:
            self.display_text = "{0}\n{1}".format(self.source_text, self.target_text)
        return self

File: src/frontend/src/types/subtitle.ts
export interface SubtitleCue {
  index: number;
  start: number;
  end: number;
  source_text: string;
  target_text: string;
  display_text: string | null;
}

export function formatCueTime(seconds: number): string {
  const safeSeconds = Math.max(0, seconds);
  const minutes = Math.floor(safeSeconds / 60);
  const wholeSeconds = Math.floor(safeSeconds % 60);
  const milliseconds = Math.round((safeSeconds - Math.floor(safeSeconds)) * 1000);

  return [
    String(minutes).padStart(2, "0"),
    ":",
    String(wholeSeconds).padStart(2, "0"),
    ".",
    String(milliseconds).padStart(3, "0"),
  ].join("");
}
